### PR TITLE
8277868: Use Comparable.compare() instead of surrogate code

### PR DESCRIPTION
**Pull request by stsypanov**
Instead of something like
```java
long x;
long y;
return (x < y) ? -1 : ((x == y) ? 0 : 1);
```
we can use `return Long.compare(x, y);`

All replacements are done with IDE.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277868](https://bugs.openjdk.java.net/browse/JDK-8277868): Use Comparable.compare() instead of surrogate code

### Reviewers
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to 6744a5627ed56f2e6dfaf071dd9aa86365457e23
 * [Alexey Ivanov](https://openjdk.java.net/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6575/head:pull/6575` \
`$ git checkout pull/6575`

Update a local copy of the PR: \
`$ git checkout pull/6575` \
`$ git pull https://git.openjdk.java.net/jdk pull/6575/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6575`

View PR using the GUI difftool: \
`$ git pr show -t 6575`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6575.diff">https://git.openjdk.java.net/jdk/pull/6575.diff</a>

</details>

### CODE REVIEW

**Review by dfuch — COMMENTED** [review]
Changes to java.net look good. Please obtain approval from reviewers in the other areas before integrating.

File: src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java [hunk 1]
```diff
@@ -251,7 +246,7 @@ private void moveWithinTableRange(JTable table, int dx, int dy) {
         }
 
         private static int sign(int num) {
-            return (num < 0) ? -1 : ((num == 0) ? 0 : 1);
+            return Integer.compare(num, 0);
```
Comment (mbien) [inline src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java:249]: => Integer.signum(num)

Comment (stsypanov) [reply src/java.desktop/share/classes/javax/swing/plaf/basic/BasicTableUI.java]: Done!

**Review by RogerRiggs — APPROVED** [review]
The core-libs file changes look fine.
A 'client' reviewer should take a look too.

File: src/java.desktop/share/classes/java/awt/geom/Line2D.java [hunk 2]
```diff
@@ -112,15 +112,15 @@ public Float(Point2D p1, Point2D p2) {
          * @since 1.2
          */
         public double getX1() {
-            return (double) x1;
+            return x1;
```
Comment (prrace) [inline src/java.desktop/share/classes/java/awt/geom/Line2D.java:115]: What do these changes have to do with the subject of the PR ?

File: src/java.desktop/share/classes/sun/java2d/Spans.java [hunk 3]
```diff
@@ -315,19 +315,11 @@ boolean contains(float pos) {
...
-            if (mStart < otherStart) {
-                result = -1;
-            } else if (mStart > otherStart) {
-                result = 1;
-            } else {
-                result = 0;
-            }
 
-            return result;
+            return Float.compare(mStart, otherStart);
```
Comment (prrace) [inline src/java.desktop/share/classes/sun/java2d/Spans.java:322]: We don't need the variable any more, do we ?

File: src/java.desktop/share/classes/java/awt/geom/Line2D.java (hunk 2 above)
Comment (stsypanov) [reply src/java.desktop/share/classes/java/awt/geom/Line2D.java:115]: Just a tiny clean-up in on of affected files. Do you want this to be reverted?

File: src/java.desktop/share/classes/sun/java2d/Spans.java (hunk 3 above)
Comment (stsypanov) [reply src/java.desktop/share/classes/sun/java2d/Spans.java]: inlined

File: src/java.base/share/classes/java/util/Calendar.java [hunk 4]
```diff
@@ -3417,7 +3417,7 @@ private void updateTime() {
 
     private int compareTo(long t) {
         long thisTime = getMillisOf(this);
-        return (thisTime > t) ? 1 : (thisTime == t) ? 0 : -1;
+        return Long.compare(thisTime, t);
```
Comment (aivanov-jdk) [inline src/java.base/share/classes/java/util/Calendar.java:3420]: Probably, in this case `thisTime` variable can also be dropped.

File: src/java.base/share/classes/java/util/Date.java [hunk 5]
```diff
@@ -975,10 +970,11 @@ static final long getMillisOf(Date date) {
      * @since   1.2
      * @throws    NullPointerException if {@code anotherDate} is null.
      */
+    @Override
     public int compareTo(Date anotherDate) {
         long thisTime = getMillisOf(this);
         long anotherTime = getMillisOf(anotherDate);
-        return (thisTime<anotherTime ? -1 : (thisTime==anotherTime ? 0 : 1));
+        return Long.compare(thisTime, anotherTime);
```
Comment (aivanov-jdk) [inline src/java.base/share/classes/java/util/Date.java:977]: Looks like local variables can also be dropped here as each value is used once.

File: src/java.base/share/classes/java/util/UUID.java [hunk 6]
```diff
@@ -514,8 +514,6 @@ public int compareTo(UUID val) {
         // can simply be numerically compared as two numbers
         return (this.mostSigBits < val.mostSigBits ? -1 :
                 (this.mostSigBits > val.mostSigBits ? 1 :
-                 (this.leastSigBits < val.leastSigBits ? -1 :
-                  (this.leastSigBits > val.leastSigBits ? 1 :
-                   0))));
+                 Long.compare(this.leastSigBits, val.leastSigBits)));
```
Comment (aivanov-jdk) [inline src/java.base/share/classes/java/util/UUID.java:517]: In this case Javadoc specifies only -1, 0 or 1 can be returned. `Long.compare` does not specify this but returns these values. Couldn't it cause any problems in the future if implementation of `Long.compare` is changed?

Does it make sense to use `Long.compare` for `mostSigBits` too?
```java
int mostSigBits = Long.compare(this.mostSigBits, val.mostSigBits);
return mostSigBits != 0 ? mostSigBits : Long.compare(this.leastSigBits, val.leastSigBits);
```

File: src/java.desktop/share/classes/java/awt/geom/Line2D.java (hunk 2 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/java/awt/geom/Line2D.java:115]: I believe it makes to preserve the cast: the fields are of type `float` and explicit cast hints there's a type conversion.

File: src/java.base/share/classes/java/util/Calendar.java (hunk 4 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/util/Calendar.java:3420]: Inlined

File: src/java.base/share/classes/java/util/Date.java (hunk 5 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/util/Date.java:977]: Inlined

File: src/java.desktop/share/classes/java/awt/geom/Line2D.java (hunk 2 above)
Comment (stsypanov) [reply src/java.desktop/share/classes/java/awt/geom/Line2D.java]: Reverted

File: src/java.base/share/classes/java/util/UUID.java (hunk 6 above)
Comment (stsypanov) [reply src/java.base/share/classes/java/util/UUID.java:517]: This behavior is specified in `Comparator` itself, so is hardly to be ever changed, I think

Comment (aivanov-jdk) [reply src/java.base/share/classes/java/util/UUID.java:517]: > This behavior is specified in `Comparator` itself, so is hardly to be ever changed, I think

Not quite. [`Comparator.compare` returns](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Comparator.html#compare(T,T)): “a negative integer, zero, or a positive integer as the first argument is less than, equal to, or greater than the second.” [`Comparable.compareTo`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Comparable.html#compareTo(T)) does the same. Yet Javadoc for `UUID` specifically requires `-1`, `0` or `1`, which is stricter. However, the implementation of `Long.compare` complies with this requirement even though it's not specified.

Probably, it makes sense to update the Javadoc for `UUID` and relax the requirement to “a negative integer, zero, or a positive integer.”

Comment (stsypanov) [reply src/java.base/share/classes/java/util/UUID.java:517]: Done

Comment (aivanov-jdk) [reply src/java.base/share/classes/java/util/UUID.java:517]: I guess I didn't make myself clear. It wasn't a showstopper but it's still an issue.

This kind of change to Javadoc requires a CSR therefore it's better to separate it from this changeset which touches other files.

I see two options: 1) leave this unchanged and update implementation together with Javadoc; or 2) integrate the update to the implementation which you already have and update the Javadoc in a follow-up bug.

Comment (aivanov-jdk) [reply src/java.base/share/classes/java/util/UUID.java:517]: The suggestion to compare `mostSigBits` using `Long.compare` if it looks reasonable still applies.

Comment (stsypanov) [reply src/java.base/share/classes/java/util/UUID.java:517]: Ok, let me revert JavaDoc and apply the suggestion

**Review by aivanov-jdk — APPROVED** [review]

**Comment by RogerRiggs** [comment]
As it turns out replacing the code was not 100% equivalent and a test failure resulted.
See https://bugs.openjdk.java.net/browse/JDK-8278937

Double.compare and the original code handle the non-numeric forms of Double differently.

**Pull request closed without merging**